File: acmicpc/1477.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
)

var bufin = bufio.NewReader(os.Stdin)
var bufout = bufio.NewWriter(os.Stdin)
var n, m, l, k int
var num []int

func check(mid int) bool {
	ret := 0
	for i:=0; i<len(num)-1; i++ {
		ret += (num[i+1] - num[i] - 1) / mid;
	}
	return ret > m
}

func main() {
	defer bufout.Flush()
	fmt.Fscan(bufin, &n, &m, &l)

	for i:=0; i<n; i++ {
		fmt.Fscan(bufin, &k)
		num = append(num, k)
	}
	num = append(num, 0)
	num = append(num, l)

	sort.Ints(num)

	x := 0
	y := l
	for {
		if x > y {
			break
		}
		mid := (x+y)/2
		if check(mid) {
			x = mid+1;
		} else {
			y = mid-1;
		}
	}
	fmt.Println(x)
}

File: acmicpc/1477.cpp
#include <iostream>
#include <algorithm>
#include <vector>
using namespace std;

int n, m, l, k;
vector<int> vec;

bool check(int mid)
{
	int ret = 0;
	for (int i=0; i<vec.size()-1; ++i)
		ret += (vec[i+1] - vec[i] - 1) / mid;

	return ret > m;
}

int main()
{
	cin >> n >> m >> l;
	vec.push_back(0);
	for (int i=0; i<n; ++i) {
		cin >> k;
		vec.push_back(k);
	}
	vec.push_back(l);

	sort(vec.begin(), vec.end());

	int x = 0, y = l;
	while (x <= y) {
		int mid = (x+y)/2;
		if (check(mid))
			x = mid+1;
		else
			y = mid-1;
	}
	cout << x << '\n';

	return 0;
}
